Drop support for .Net Standard 2.1

batiati opened the pull request:
In practice, drops the support for .Net Core 3.1 (already in [EOL](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core)) and .Net <= 6 (in [EOL](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core) in Dec 2024). The minimum supported version now is .Net 7.0; Recommended version is .Net 8.0 LTS.

- Simplifies the code, removing the custom `UInt128` type, and version-specific conditionals (i.e `#if NETSTANDARD` and `#if !NET7_0_OR_GREATER`).

- Removes the burden of publishing dual-version artifacts on Nuget.

- Speeds up the CI by not having to run the same tests for each supported version.

- Bonus: Introduces ~top-level~ file-scoped namespaces, saving one level of indentation.

matklad commented:
> Bonus: Introduces top-level namespaces, saving one level of indentation.

Next time, would be great to file indentation/formatting changes as a separate commit, 3k diff is a bit intimidating ;-)

matklad commented on src/clients/dotnet/TigerBeetle.Tests/ExceptionTests.cs:
```diff
@@ -1,81 +1,81 @@
+    [TestMethod]
+    public void AssertTrue()
+    {
+        AssertionException.AssertTrue(1 == 1);
```
In a separate PR, could you add a documentation for `AssertionException`? This class, and these tests in particular, look quiet peculiar, as they seem tautological --- we we aren't just using `Assert`, why we need our own `AssertionException`? 

My _guess_ here is that the built-in assertion mechanism in .Net is off by default, so we need our own asserts that are guaranteed to always through, but I don't really know. Would be great to have that documented for future readers!

matklad commented on src/clients/dotnet/TigerBeetle.Tests/UInt128Tests.cs:
```diff
@@ -4,149 +4,102 @@
+
+
+    [TestMethod]
+    public void BigIntegerConvertion()
```
In a separate PR, let's also add tests for small numbers, like `0`, and also a test for a _negative_ BigInteger (which we expect to throw)

matklad reviewed (dismissed):
LGTM!

I haven't reviewed every line, and I don't know .Net enough to review for unknown-unknowns anyway, but this all reads like a clear improvement, thanks batiati!

batiati committed:
Drop support for .Net 6

- Remove NetStandard21 from csproj;

- Removes preprocessor's conditions for NETSTANDARD;

- Removes the custom type UInt128 and the preprocessor condition NET7_0_OR_GREATER;

- Updates the CI and docs with the new supported version;

- Changes dotnet_bindings.zig to render code using top-level namespaces.

- Adds tests for converting BigInteger to UInt128 using zero and negative numbers;

- Adds documentation for AssertionException;

batiati committed:
File-scoped namespaces, removing one indentation level.

batiati commented:
> would be great to file indentation/formatting changes as a separate commit, 3k diff is a bit intimidating ;-)

It's the last day of the year! Let's do it right now 😂
Also, the review's comments were addressed in this PR.

matklad commented on src/clients/dotnet/TigerBeetle/AssertionException.cs:
```diff
@@ -2,6 +2,12 @@
+    /// Correctness check assertions differ from other assertions offered by the .Net ecosystem
+    /// i.e. Tracer.Assert and Debug.Assert, because they are not meant to be disabled or re-routed.
+    /// It's recommended that the application handle AssertionException as unrecoverable fatal errors.
+    /// </remarks>
```
Opening tag (summary) and closing tag (remarks) do not match?

batiati replied:
Nice catch, surely they don't match 😳

Important thing, this exposed that we are not generating the XML documentation file. Otherwise, it should be an error (or maybe a warning) during the build.

batiati committed:
Fix malformed XML comment

batiati committed:
Malformed XML comments are just warnings. As per the NASA rules, we shouldn't ignore warnings!

- Generate the XML documentation file used by IDEs.
- Treat warnings as errors.
- Ignore missing XML comments for public members (we don't provide documentation for all of them)

matklad commented on src/clients/dotnet/TigerBeetle/TigerBeetle.csproj:
```diff
@@ -9,6 +9,9 @@
     <RollForward>LatestMajor</RollForward>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <NoWarn>1591</NoWarn> <!-- Disables the warning CS1591: Missing XML comment for publicly visible type or member -->
```
:heart:
